Source/WebKit/UIProcess/WebPageProxy.cpp

```diff
@@ -3254,65 +3257,62 @@ void WebPageProxy::handlePreventableTouchEvent(NativeWebTouchEvent& event)
         didReceiveEvent(event.type(), false);
         if (isTouchStart)
             pageClient().doneDeferringTouchStart(false);
+        if (isTouchMove)
+            pageClient().doneDeferringTouchMove(false);
         if (isTouchEnd)
             pageClient().doneDeferringTouchEnd(false);
         return;
     }
 
-    if (isTouchStart || isTouchEnd) {
-        if (isTouchStart)
-            ++m_handlingPreventableTouchStartCount;
+    if (isTouchStart)
+        ++m_handlingPreventableTouchStartCount;
 
-        if (isTouchEnd)
-            ++m_handlingPreventableTouchEndCount;
+    if (isTouchMove) {
+        if (!m_handlingPreventableTouchMoveCount)
+            m_hasPreventedTouchMove = false;
+        ++m_handlingPreventableTouchMoveCount;
+    }
 
-        sendWithAsyncReply(Messages::EventDispatcher::TouchEvent(m_webPageID, event), [this, weakThis = WeakPtr { *this }, event] (bool handled) {
-            RefPtr protectedThis { weakThis.get() };
-            if (!protectedThis)
-                return;
+    if (isTouchEnd)
+        ++m_handlingPreventableTouchEndCount;
 
-            bool didFinishDeferringTouchStart = false;
-            ASSERT_IMPLIES(event.type() == WebEvent::TouchStart, m_handlingPreventableTouchStartCount);
-            if (event.type() == WebEvent::TouchStart && m_handlingPreventableTouchStartCount)
-                didFinishDeferringTouchStart = !--m_handlingPreventableTouchStartCount;
+    sendWithAsyncReply(Messages::EventDispatcher::TouchEvent(m_webPageID, event), [this, weakThis = WeakPtr { *this }, event] (bool handled) {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis)
+            return;
 
-            bool didFinishDeferringTouchEnd = false;
-            ASSERT_IMPLIES(event.type() == WebEvent::TouchEnd, m_handlingPreventableTouchEndCount);
-            if (event.type() == WebEvent::TouchEnd && m_handlingPreventableTouchEndCount)
-                didFinishDeferringTouchEnd = !--m_handlingPreventableTouchEndCount;
+        bool didFinishDeferringTouchStart = false;
+        ASSERT_IMPLIES(event.type() == WebEvent::TouchStart, m_handlingPreventableTouchStartCount);
+        if (event.type() == WebEvent::TouchStart && m_handlingPreventableTouchStartCount)
+            didFinishDeferringTouchStart = !--m_handlingPreventableTouchStartCount;
 
-            bool handledOrFailedWithError = handled || m_handledSynchronousTouchEventWhileDispatchingPreventableTouchStart;
-            if (!isHandlingPreventableTouchStart())
-                m_handledSynchronousTouchEventWhileDispatchingPreventableTouchStart = false;
+        bool didFinishDeferringTouchMove = false;
+        ASSERT_IMPLIES(event.type() == WebEvent::TouchMove, m_handlingPreventableTouchMoveCount);
+        if (event.type() == WebEvent::TouchMove && m_handlingPreventableTouchMoveCount) {
+            m_hasPreventedTouchMove |= handled;
```

Comment:
Should we maybe make `m_hasPreventedTouchMove` into an `std::optional<bool>` so that we only ever set it on the first `"touchmove"`?  I think it might be better to have things be (slightly) more predictable and (slightly) list racey.

Reply:
I suppose we can....try. Not sure if it's worth the risk, though.